src/main/scala/io/iohk/ethereum/ledger/BranchResolution.scala

```diff
@@ -37,52 +37,32 @@ class BranchResolution(blockchain: Blockchain) {
     val oldBlocks = oldBlocksWithCommonPrefix.drop(commonPrefixLength)
     val newHeaders = headersList.drop(commonPrefixLength)
 
-    if (compareByCheckpoints(newHeaders, oldBlocks.map(_.header)))
-      NewBetterBranch(oldBlocks)
-    else
-      NoChainSwitch
-  }
-
-  /**
-    * @return true if newBranch is better than oldBranch
-    */
-  private def compareByCheckpoints(newBranch: Seq[BlockHeader], oldBranch: Seq[BlockHeader]): Boolean =
-    (branchLatestCheckpoint(newBranch), branchLatestCheckpoint(oldBranch)) match {
-      case (Some(newCheckpoint), Some(oldCheckpoint)) =>
-        if (newCheckpoint.number == oldCheckpoint.number)
-          compareByDifficulty(newBranch, oldBranch)
+    val maybeParentWeight =
+      oldBlocks.headOption
+        .map(_.header.parentHash)
+        .orElse(newHeaders.headOption.map(_.parentHash))
+        .flatMap(blockchain.getChainWeightByHash)
+
+    maybeParentWeight match {
+      case Some(parentWeight) =>
+        val oldWeight = oldBlocks.foldLeft(parentWeight)((w, b) => w.increase(b.header))
+        val newWeight = newHeaders.foldLeft(parentWeight)((w, h) => w.increase(h))
+
+        if (newWeight > oldWeight)
+          NewBetterBranch(oldBlocks)
         else
-          newCheckpoint.number > oldCheckpoint.number
-
-      case (Some(_), None) =>
-        true
-
-      case (None, Some(_)) =>
-        false
+          NoChainSwitch
 
-      case (None, None) =>
-        compareByDifficulty(newBranch, oldBranch)
+      case None =>
+        // TODO: should we log if parentWeight not found?
```

Comment:
why not

Reply:
Cause the code would be ugly :) This `match` doesn't capture the situation - there could be no parent to get the weight for